File: built-in plugins dsp/mb-comp-parallel.cpp
/**
 * MB Parallel Comp
 * Category : effect
 * Type     : compressor
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : NY-style parallel compression
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_COMP_PARALLEL_H
#define MB_COMP_PARALLEL_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbCompParallel : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-comp-parallel";
    static constexpr const char* PLUGIN_NAME    = "MB Parallel Comp";
    static constexpr const char* PLUGIN_TYPE    = "compressor";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float threshold = -35f;  // range [-60, 0]
    float ratio = 8f;  // range [1, 20]
    float blend = 0.4f;  // range [0, 1]
    };

    MbCompParallel() = default;
    ~MbCompParallel() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.threshold = std::clamp(params.threshold, -60f, 0f);
        params.ratio = std::clamp(params.ratio, 1f, 20f);
        params.blend = std::clamp(params.blend, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Parallel Comp
        return input;
    }
};

#endif // MB_COMP_PARALLEL_H

File: built-in plugins dsp/mb-comp-parallel.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbCompParallelPlugin: PluginDefinition = {
  id: "mb-comp-parallel",
  slug: "mb-comp-parallel",
  name: "MB Parallel Comp",
  category: "effect",
  type: "compressor",
  version: "1.0.0",
  description: "NY-style parallel compression",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "threshold",
      name: "Threshold",
      type: "float",
      defaultValue: -35,
      minValue: -60,
      maxValue: 0,
      automatable: true,
    },
    {
      id: "ratio",
      name: "Ratio",
      type: "float",
      defaultValue: 8,
      minValue: 1,
      maxValue: 20,
      automatable: true,
    },
    {
      id: "blend",
      name: "Blend",
      type: "float",
      defaultValue: 0.4,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { threshold: -35, ratio: 8, blend: 0.4 },
};

export default MbCompParallelPlugin;
